File: scripts/radio_binary_codec.py
"""Standalone v1 codecs for the planned binary radio stores."""

import struct

MAGIC_CATALOG = b"RDBC"
MAGIC_FAVORITES = b"RFAV"
VERSION = 1
MAX_RECORDS = 128
MAX_UUID = 64
FIELD_LIMITS = {
    "stationuuid": 64,
    "name": 256,
    "url_resolved": 512,
    "codec": 32,
    "country": 128,
    "countrycode": 16,
    "state": 128,
    "language": 64,
    "tags": 512,
}
HEADER_SIZE = 16
MAX_RECORD_LENGTH = 2052


class BinaryFormatError(ValueError):
    pass


def encode_header(magic: bytes, count: int) -> bytes:
    if magic not in (MAGIC_CATALOG, MAGIC_FAVORITES):
        raise BinaryFormatError("invalid magic")
    if not 0 <= count <= MAX_RECORDS:
        raise BinaryFormatError("invalid record count")
    return magic + struct.pack("<HHII", VERSION, 0, count, 0)


def decode_header(data: bytes, magic: bytes) -> int:
    if len(data) < HEADER_SIZE:
        raise BinaryFormatError("truncated header")
    if data[:4] != magic:
        raise BinaryFormatError("bad magic")
    version, flags, count, reserved = struct.unpack_from("<HHII", data, 4)
    if version != VERSION:
        raise BinaryFormatError("unsupported version")
    if flags != 0 or reserved != 0:
        raise BinaryFormatError("invalid flags")
    if count > MAX_RECORDS:
        raise BinaryFormatError("invalid record count")
    return count


def _encode_string(value: str, limit: int) -> bytes:
    raw = value.encode("utf-8")
    if len(raw) > limit:
        raise BinaryFormatError("field too long")
    return struct.pack("<H", len(raw)) + raw


def _decode_string(data: bytes, offset: int, end: int, limit: int):
    if offset + 2 > end:
        raise BinaryFormatError("truncated field length")
    length = struct.unpack_from("<H", data, offset)[0]
    offset += 2
    if length > limit or length > end - offset:
        raise BinaryFormatError("invalid field length")
    raw = data[offset:offset + length]
    return raw.decode("utf-8"), offset + length


def encode_catalog_record(station: dict) -> bytes:
    payload = bytearray()
    for field in ("stationuuid", "name", "url_resolved", "codec"):
        payload += _encode_string(station.get(field, ""), FIELD_LIMITS[field])
    payload += struct.pack("<I", station.get("bitrate", 0))
    for field in ("country", "countrycode", "state", "language", "tags"):
        payload += _encode_string(station.get(field, ""), FIELD_LIMITS[field])
    total = 4 + len(payload)
    if total > MAX_RECORD_LENGTH:
        raise BinaryFormatError("record too long")
    return struct.pack("<I", total) + payload


def decode_catalog_record(data: bytes) -> dict:
    if len(data) < 4:
        raise BinaryFormatError("truncated record")
    total = struct.unpack_from("<I", data)[0]
    if total < 4 or total > MAX_RECORD_LENGTH or total != len(data):
        raise BinaryFormatError("invalid record length")
    offset, end = 4, total
    station = {}
    for field in ("stationuuid", "name", "url_resolved", "codec"):
        station[field], offset = _decode_string(data, offset, end, FIELD_LIMITS[field])
    if offset + 4 > end:
        raise BinaryFormatError("truncated bitrate")
    station["bitrate"] = struct.unpack_from("<I", data, offset)[0]
    offset += 4
    for field in ("country", "countrycode", "state", "language", "tags"):
        station[field], offset = _decode_string(data, offset, end, FIELD_LIMITS[field])
    if offset != end:
        raise BinaryFormatError("unexpected record bytes")
    return station


def encode_favorite_uuid(uuid: str) -> bytes:
    raw = uuid.encode("utf-8")
    if not raw or len(raw) > MAX_UUID:
        raise BinaryFormatError("invalid UUID")
    return struct.pack("<H", len(raw)) + raw


def decode_favorite_uuid(data: bytes) -> str:
    if len(data) < 2:
        raise BinaryFormatError("truncated UUID length")
    length = struct.unpack_from("<H", data)[0]
    if not 0 < length <= MAX_UUID or length + 2 != len(data):
        raise BinaryFormatError("invalid UUID record")
    try:
        return data[2:].decode("utf-8")
    except UnicodeDecodeError as exc:
        raise BinaryFormatError("invalid UUID UTF-8") from exc


def decode_catalog_file(data: bytes):
    count = decode_header(data[:HEADER_SIZE], MAGIC_CATALOG)
    offset, records = HEADER_SIZE, []
    for _ in range(count):
        if offset + 4 > len(data):
            raise BinaryFormatError("truncated record")
        total = struct.unpack_from("<I", data, offset)[0]
        if total > len(data) - offset:
            raise BinaryFormatError("truncated record")
        records.append(decode_catalog_record(data[offset:offset + total]))
        offset += total
    if offset != len(data):
        raise BinaryFormatError("unexpected trailing bytes")
    return records


def decode_favorites_file(data: bytes):
    count = decode_header(data[:HEADER_SIZE], MAGIC_FAVORITES)
    offset, records = HEADER_SIZE, []
    for _ in range(count):
        if offset + 2 > len(data):
            raise BinaryFormatError("truncated UUID length")
        length = struct.unpack_from("<H", data, offset)[0]
        end = offset + 2 + length
        if end > len(data):
            raise BinaryFormatError("truncated UUID record")
        records.append(decode_favorite_uuid(data[offset:end]))
        offset = end
    if offset != len(data):
        raise BinaryFormatError("unexpected trailing bytes")
    return records

File: main/media/radio_binary_codec.cc
#include "radio_binary_codec.h"

#include <array>
#include <cstring>

namespace RadioBinaryCodec {
namespace {
constexpr size_t kHeaderSize = 16;
constexpr std::array<uint16_t, 9> kLimits = {64, 256, 512, 32, 128, 16, 128, 64, 512};
constexpr char kCatalogMagic[] = "RDBC";
constexpr char kFavoritesMagic[] = "RFAV";

void Put16(std::string& out, uint16_t value) {
    out.push_back(static_cast<char>(value));
    out.push_back(static_cast<char>(value >> 8));
}
void Put32(std::string& out, uint32_t value) {
    for (int i = 0; i < 4; ++i) out.push_back(static_cast<char>(value >> (i * 8)));
}
bool Get16(const uint8_t* data, size_t size, size_t& pos, uint16_t& value) {
    if (pos > size || size - pos < 2) return false;
    value = static_cast<uint16_t>(data[pos]) | (static_cast<uint16_t>(data[pos + 1]) << 8);
    pos += 2;
    return true;
}
bool Get32(const uint8_t* data, size_t size, size_t& pos, uint32_t& value) {
    if (pos > size || size - pos < 4) return false;
    value = static_cast<uint32_t>(data[pos]) | (static_cast<uint32_t>(data[pos + 1]) << 8) |
            (static_cast<uint32_t>(data[pos + 2]) << 16) |
            (static_cast<uint32_t>(data[pos + 3]) << 24);
    pos += 4;
    return true;
}
bool Header(const char* magic, uint32_t count, std::string& out) {
    if (count > kMaxRecords) return false;
    out.assign(magic, 4);
    Put16(out, 1); Put16(out, 0); Put32(out, count); Put32(out, 0);
    return true;
}
bool ParseHeader(const char* magic, const uint8_t* data, size_t size, uint32_t& count) {
    if (data == nullptr || size < kHeaderSize || std::memcmp(data, magic, 4) != 0) return false;
    size_t pos = 4; uint16_t version, flags; uint32_t reserved;
    if (!Get16(data, size, pos, version) || !Get16(data, size, pos, flags) ||
        !Get32(data, size, pos, count) || !Get32(data, size, pos, reserved)) return false;
    return version == 1 && flags == 0 && reserved == 0 && count <= kMaxRecords;
}
bool PutString(std::string& out, const std::string& value, uint16_t limit) {
    if (value.size() > limit) return false;
    Put16(out, static_cast<uint16_t>(value.size())); out.append(value); return true;
}
bool GetString(const uint8_t* data, size_t size, size_t& pos, uint16_t limit, std::string& value) {
    uint16_t length;
    if (!Get16(data, size, pos, length) || length > limit || pos > size || size - pos < length) return false;
    value.assign(reinterpret_cast<const char*>(data + pos), length); pos += length; return true;
}
}  // namespace

bool EncodeCatalogHeader(uint32_t count, std::string& out) { return Header(kCatalogMagic, count, out); }
bool DecodeCatalogHeader(const uint8_t* data, size_t size, uint32_t& count) { return ParseHeader(kCatalogMagic, data, size, count); }
bool EncodeFavoritesHeader(uint32_t count, std::string& out) { return Header(kFavoritesMagic, count, out); }
bool DecodeFavoritesHeader(const uint8_t* data, size_t size, uint32_t& count) { return ParseHeader(kFavoritesMagic, data, size, count); }

bool EncodeCatalogRecord(const RadioStationInfo& station, std::string& out) {
    std::string payload;
    const std::array<const std::string*, 4> first = {&station.stationuuid, &station.name, &station.url_resolved, &station.codec};
    for (size_t i = 0; i < first.size(); ++i) if (!PutString(payload, *first[i], kLimits[i])) return false;
    Put32(payload, station.bitrate);
    const std::array<const std::string*, 5> last = {&station.country, &station.countrycode, &station.state, &station.language, &station.tags};
    for (size_t i = 0; i < last.size(); ++i) if (!PutString(payload, *last[i], kLimits[i + 4])) return false;
    if (payload.size() > kMaxRecordLength - 4) return false;
    out.clear(); Put32(out, static_cast<uint32_t>(payload.size() + 4)); out += payload; return true;
}

bool DecodeCatalogRecord(const uint8_t* data, size_t size, RadioStationInfo& station) {
    if (data == nullptr || size < 4 || size > kMaxRecordLength) return false;
    size_t pos = 0; uint32_t total;
    if (!Get32(data, size, pos, total) || total != size || total < 4) return false;
    const std::array<std::string*, 4> first = {&station.stationuuid, &station.name, &station.url_resolved, &station.codec};
    for (size_t i = 0; i < first.size(); ++i) if (!GetString(data, size, pos, kLimits[i], *first[i])) return false;
    if (!Get32(data, size, pos, station.bitrate)) return false;
    const std::array<std::string*, 5> last = {&station.country, &station.countrycode, &station.state, &station.language, &station.tags};
    for (size_t i = 0; i < last.size(); ++i) if (!GetString(data, size, pos, kLimits[i + 4], *last[i])) return false;
    return pos == size;
}

bool EncodeFavoriteUuid(const std::string& uuid, std::string& out) {
    if (uuid.empty() || uuid.size() > kMaxUuidLength) return false;
    out.clear(); Put16(out, static_cast<uint16_t>(uuid.size())); out += uuid; return true;
}
bool DecodeFavoriteUuid(const uint8_t* data, size_t size, std::string& uuid) {
    size_t pos = 0; uint16_t length;
    if (data == nullptr || !Get16(data, size, pos, length) || length == 0 || length > kMaxUuidLength || size - pos != length) return false;
    uuid.assign(reinterpret_cast<const char*>(data + pos), length); return true;
}
}  // namespace RadioBinaryCodec
